### ISSUE TITLE
生成对抗样本，梯度不更新，报错AttributeError: 'Tensor' object has no attribute 'trainable'

### ISSUE THREAD

**Issue by helen88** [PaddlePaddle/Paddle#31798]
环境：paddlepaddle2.0
我想要更新图像数据的梯度，
img = paddle.to_tensor(img, dtype='float32', place=paddle.CUDAPlace(0), stop_gradient=False)
model = paddle.vision.models.vgg16(pretrained=True)
predict = model(img)[0]
print (predict.shape)
label = np.argmax(predict)

for param in model.parameters():
    param.stop_gradient = True

#optimizer = torch.optim.Adam([img])

optimizer = paddle.optimizer.Adam(learning_rate=0.001, parameters=img)

epochs=100
target=288
target = paddle.to_tensor(target, dtype='int64',place=paddle.CUDAPlace(0))

for epoch in range(epochs):
    # forward + backward
    output = model(img)
   
    loss = F.cross_entropy(output, target)
    label = np.argmax(output[0])
    #print("label={}".format(label))
    print("epoch={} loss={} label={}".format(epoch,loss,label))

    #如果定向攻击成功
    if label == target:
        break
    loss.backward()
    optimizer.step()
    optimizer.clear_grad()
 运行报错：不确定是优化器的parameters参数设置有问题，还是用paddle.to_tensor创建的img有问题。麻烦帮忙看下，怎么才能更新到img。
![image](https://user-images.githubusercontent.com/4593091/112095282-2df20f80-8bd7-11eb-9816-6ac4c877746d.png)

**Comment by JZ-LIANG** [comment]
param.stop_gradient = True 这句应该去掉

param.stop_gradient = True 的意思是： 从这个param往input 追溯，所以有在该param 之前的grad 都不会计算。
https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/faq/train_cn.html#stop-gradient-true

adv sample 需要求对input 的grad， 所以模型中间每一层的grad 都需要在 “计算 input grad 前” 被计算。

**Comment by helen88** [comment]
optimizer = paddle.optimizer.Adam(learning_rate=0.001, parameters=img) 报错是这里的“parameters=img”不能接受一个tensor类型，我想要更新我输入图片的梯度，这里应该怎么定义呢。

**Comment by helen88** [comment]
> https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/faq/train_cn.html#stop-gradient-true
去掉还是一样的报错：
_AttributeError: 'Tensor' object has no attribute 'trainable'_。
另外：adv sample是要求模型参数不更新，只对input更新，这里需要怎么把input image 定义成可进行梯度更新的变量，用tensor不可以。
1.  img = paddle.to_tensor(img, dtype='float32', place=paddle.CUDAPlace(0), stop_gradient=False) 这么定义就会报错。
2. optimizer = paddle.optimizer.Adam(learning_rate=0.001, parameters=img) 这个写法没报错，但optimizer.step()就会报错了

**Comment by JZ-LIANG** [comment]
> > https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/faq/train_cn.html#stop-gradient-true
> > 去掉还是一样的报错：
> > _AttributeError: 'Tensor' object has no attribute 'trainable'_。
> > 另外：adv sample是要求模型参数不更新，只对input更新，这里需要怎么把input image 定义成可进行梯度更新的变量，用tensor不可以。
> 
> 1. img = paddle.to_tensor(img, dtype='float32', place=paddle.CUDAPlace(0), stop_gradient=False) 这么定义就会报错。
> 2. optimizer = paddle.optimizer.Adam(learning_rate=0.001, parameters=img) 这个写法没报错，但optimizer.step()就会报错了

参考一下这个
https://www.paddlepaddle.org.cn/documentation/docs/zh/1.8/user_guides/cv_case/gan/README.cn.html#shengchengduikangwangluo

**Comment by helen88** [comment]
> https://www.paddlepaddle.org.cn/documentation/docs/zh/1.8/user_guides/cv_case/gan/README.cn.html#shengchengduikangwangluo

我是想用paddle2.0的新特性来实现，paddle1.X版本的已经实现了，2.0要取到梯度信息都是静态图的写法，那还是静态图了，就没有利用起paddle2 动静统一的特性了。

**Comment by helen88** [comment]
update paddle, newer paddle will support Tensor get grad in training process.
